src/providers/azure/azure.go

```diff
@@ -0,0 +1,182 @@
+package azure
+
+import (
+	"bufio"
+	"encoding/xml"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/coreos-metadata/src/config"
+	"github.com/coreos/coreos-metadata/src/retry"
+)
+
+const (
+	AgentName             = "com.coreos.metadata"
+	FabricProtocolVersion = "2012-11-30"
+)
+
+func FetchMetadata() (config.Metadata, error) {
+	addr, err := getFabricAddress()
+	if err != nil {
+		return config.Metadata{}, err
+	}
+
+	if err := assertFabricCompatible(addr, FabricProtocolVersion); err != nil {
+		return config.Metadata{}, err
+	}
+
+	return fetchSharedConfig(addr)
+}
+
+func getFabricAddress() (net.IP, error) {
+	lease, err := os.Open("/run/systemd/netif/leases/2")
+	if err != nil {
+		return nil, err
+	}
+
+	line := bufio.NewScanner(lease)
+
+	var rawEndpoint string
+	for line.Scan() {
+		parts := strings.Split(line.Text(), "=")
+		if parts[0] == "OPTION_245" && len(parts) == 2 {
+			rawEndpoint = parts[1]
+			break
+		}
+	}
+
+	if len(rawEndpoint) == 0 {
+		return nil, fmt.Errorf("DHCP option 245 not found in lease")
+	}
+
+	octets := make([]byte, 4)
+	for i := 0; i < 4; i++ {
+		octet, err := strconv.ParseUint(rawEndpoint[2*i:2*i+2], 16, 8)
+		if err != nil {
+			return nil, err
+		}
+		octets[i] = byte(octet)
+	}
+
+	return net.IPv4(octets[0], octets[1], octets[2], octets[3]), nil
+}
+
+func assertFabricCompatible(endpoint net.IP, desiredVersion string) error {
+	body, err := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {AgentName},
+			"x-ms-version":    {FabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}.Get(fmt.Sprintf("http://%s/?comp=versions", endpoint))
+	if err != nil {
+		return fmt.Errorf("failed to fetch versions: %v", err)
+	}
+
+	versions := struct {
+		Supported struct {
+			Versions []string `xml:"Version"`
+		}
+		Preferred struct {
+			Version string
+		}
+	}{}
+
+	if err := xml.Unmarshal(body, &versions); err != nil {
+		return fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	for _, version := range versions.Supported.Versions {
+		if version == desiredVersion {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("fabric version %s is not compatible", desiredVersion)
+}
+
+func fetchSharedConfig(endpoint net.IP) (config.Metadata, error) {
+	client := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {AgentName},
+			"x-ms-version":    {FabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}
```

Comment:
why are you repeating yourself?  stick this in a function which returns the populated retry.Client and reuse with above, they're doing identical things.